File: cpp/examples/carter_registration.cpp
#include "registration_common.hpp"

int main() {
    examples::RegistrationSet set;
    for (const std::string name : {"carter1", "carter2", "carter3"}) {
        auto robot = examples::standard_robot(name, horus::core::RobotType::WHEELED, {0.82, 0.56, 0.60}, "wheeled");
        robot.configure_local_body_model("nova_carter");
        robot.remove_sensor("front_camera");
        auto camera = examples::standard_camera(name);
        camera->set_resolution({1280, 720});
        camera->set_fps(20);
        camera->set_teleop_streaming_type("webrtc");
        camera->configure_webrtc_transport({.bitrate_kbps = 2000, .framerate = 20});
        robot.add_sensor(camera);
        robot.add_sensor(std::make_shared<horus::robot::LaserScan>(
            "front_2d_lidar",
            name + "/front_2d_lidar",
            examples::topic(name, "front_2d_lidar/scan"),
            -3.14159f,
            3.14159f,
            0.005f,
            0.1f,
            30.0f,
            0.01f,
            "#6ED7FF",
            0.025f));
        set.robots.push_back(std::move(robot));
        auto dataviz = set.robots.back().create_dataviz();
        dataviz->add_robot_global_path(name, "/rviz/" + name + "/plan", "global_odom");
        dataviz->add_robot_local_path(name, "/rviz/" + name + "/local_plan", "global_odom");
        dataviz->add_robot_velocity_data(name, examples::topic(name, "chassis/odom"), "global_odom");
        dataviz->add_robot_odometry_trail(name, examples::topic(name, "chassis/odom"), "global_odom");
        set.datavizs.push_back(*dataviz);
    }
    set.datavizs.front().add_occupancy_grid("/shared_map", "map");
    return examples::register_or_fail(set);
}

File: rust/examples/carter_registration.rs
mod common;

use horus::{RobotDimensions, RobotType, WebRtcTransportConfig};
use std::sync::Arc;

fn main() {
    let mut set = common::RegistrationSet {
        robots: Vec::new(),
        datavizs: Vec::new(),
    };

    for name in ["carter1", "carter2", "carter3"] {
        let mut robot = common::standard_robot(
            name,
            RobotType::Wheeled,
            RobotDimensions::new(0.82, 0.56, 0.60),
            "wheeled",
        );
        robot.configure_local_body_model("nova_carter", true);
        robot.remove_sensor("front_camera");

        let mut camera = common::standard_camera(name);
        camera.resolution = (1280, 720);
        camera.fps = 20;
        camera.teleop_streaming_type = "webrtc".to_string();
        camera.configure_webrtc_transport(WebRtcTransportConfig {
            bitrate_kbps: Some(2000),
            framerate: Some(20),
            ..Default::default()
        });
        robot.add_sensor(Arc::new(camera)).unwrap();
        robot
            .add_sensor(Arc::new(common::standard_laser_scan(
                "front_2d_lidar",
                format!("{name}/front_2d_lidar"),
                common::topic(name, "front_2d_lidar/scan"),
            )))
            .unwrap();

        let mut dataviz = robot.create_dataviz(None);
        dataviz.add_robot_global_path(name, &format!("/rviz/{name}/plan"), "global_odom", None);
        dataviz.add_robot_local_path(
            name,
            &format!("/rviz/{name}/local_plan"),
            "global_odom",
            None,
        );
        dataviz.add_robot_velocity_data(
            name,
            &common::topic(name, "chassis/odom"),
            "global_odom",
            None,
        );
        dataviz.add_robot_odometry_trail(
            name,
            &common::topic(name, "chassis/odom"),
            "global_odom",
            None,
        );

        set.robots.push(robot);
        set.datavizs.push(dataviz);
    }

    set.datavizs[0].add_occupancy_grid("/shared_map", "map", None);
    common::register_or_fail(&mut set);
}
